sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -401,6 +401,78 @@ case class DayOfYear(child: Expression) extends UnaryExpression with ImplicitCas
   }
 }
 
+abstract class NumberToTimestampBase extends UnaryExpression
+  with ImplicitCastInputTypes {
```

Comment:
Does bigquery support things like `timestamp_seconds('123')`? I just realized that Spark has a very loose implicit cast, and support thing like `timestamp_seconds('123')` and `timestamp_seconds(123.123)`. I'm not sure if it's expected.

Reply:
It does -- it only casts string literals though, not variables. I think it's fine to allow this, given that this is Spark semantics. It *will* make it harder to add other overloads for these functions later (especially if we want string overloads that do something else), but I don't expect that to happen, so we're fine.

Reply:
This reminds me of https://github.com/apache/spark/pull/27965

In general I think we want to be conservative and don't rely on the implicit cast too much. We can support string literal as a special case later, via the framework.

For this case, how about we follow `DateAdd`? i.e.
1. implements `ExpectsInputTypes`
2. declare `IntegralType` as the input type
3. in `eval`, cast value to long via `value.asInstanceOf[Number].intValue()`
4. in codegen, cast value to long via`Math.multiplyExact((long) v, scale)`

Reply:
Seems nice,I will add it later.

Reply:
`value.asInstanceOf[Number].intValue() ` -> `value.asInstanceOf[Number].longValue()` ?